agent/src/util/device_plugin_service.rs

```diff
@@ -476,27 +543,93 @@ async fn try_update_instance_device_usage(
-            instance
-                .device_usage
-                .insert(device_usage_id.to_string(), node_name.to_string());
+        let device_usage_string = instance.device_usage.get(device_usage_id);
+        if device_usage_string.is_none() {
+            // No corresponding id found
+            trace!(
+                "try_update_instance_device_usage - could not find {} id in device_usage",
+                device_usage_id
+            );
+            return Err(Status::new(
+                Code::Unknown,
+                "Could not find device usage slot",
+            ));
+        }
 
-            if let Err(e) = kube_interface
-                .update_instance(&instance, instance_name, instance_namespace)
-                .await
-            {
-                if x == (MAX_INSTANCE_UPDATE_TRIES - 1) {
-                    trace!("try_update_instance_device_usage - update_instance returned error [{}] after max tries ... returning error", e);
-                    return Err(Status::new(Code::Unknown, "Could not update Instance"));
+        let device_usage = DeviceUsage::from_str(device_usage_string.unwrap()).map_err(|_| {
+            Status::new(
+                Code::Unknown,
+                format!(
+                    "Fails to parse {} to DeviceUsage ",
+                    device_usage_string.unwrap()
+                ),
+            )
+        })?;
+        match get_device_usage_state(&device_usage, node_name) {
+            DeviceUsageStatus::Free => {
+                let new_device_usage =
+                    DeviceUsage::create(&device_usage_kind, node_name).map_err(|e| {
+                        Status::new(
+                            Code::Unknown,
+                            format!("Fails to create DeviceUsage - {}", e),
+                        )
+                    })?;
+                instance
+                    .device_usage
+                    .insert(device_usage_id.to_string(), new_device_usage.to_string());
+
+                if let Err(e) = kube_interface
+                    .update_instance(&instance, instance_name, instance_namespace)
+                    .await
+                {
+                    if x == (MAX_INSTANCE_UPDATE_TRIES - 1) {
+                        trace!("try_update_instance_device_usage - update_instance returned error [{}] after max tries ... returning error", e);
+                        return Err(Status::new(Code::Unknown, "Could not update Instance"));
+                    }
+                    random_delay().await;
+                } else {
+                    return Ok(());
                 }
-                random_delay().await;
-            } else {
-                return Ok(());
             }
-        } else {
-            // Instance slot already reserved for this node
-            return Ok(());
-        }
+            DeviceUsageStatus::ReservedByConfiguration(_) => {
+                if matches!(device_usage_kind, DeviceUsageKind::Configuration(_)) {
```

Comment:
maybe add comment ... something like : safe to assume that this node owns the slot because get_device_usage_state only returns ReservedByConfiguration when this node owns the slot, for slots owned by other nodes (by Configuration or Instance), ReservedByOtherNode is returned.

Reply:
add comments